src/client.rs

```diff
@@ -111,18 +119,32 @@ where
     str_features: HashMap<String, CachedFeature>,
 }
 
+impl<F> CachedState<F>
+where
+    F: Enum<CachedFeature>,
+{
+    pub fn str_features<C>(&self, f: C)
+    where
+        C: FnOnce(&HashMap<String, CachedFeature>),
+    {
+        f(&self.str_features)
+    }
```

Comment:
Client.cached_state is an arc_swap, which can't be leaked. The lifetime on the alternative spelling would allow self.str_features to live for the life of the struct, but that is longer than the guard on the reference of the arc_swap, which becomes hard to reason about. To avoid that, it is easier to wrap this up in a closure form; or to return a Guard struct with a custom Deref implementation; we see the latter in other Rust APIs such as Mutex and we the former in things like https://doc.rust-lang.org/stable/std/thread/struct.LocalKey.html#method.with. I think either would work fine, but the closure approach is less code for now, which is why I chose it.

Reply:
Alternatively, perhaps I'm overthinking it and should trust the borrow checker more :P.

Reply:
Given that `arc_swap::Guard` is outside of the `CacheState` struct, the lifetime of `&self` should already be restricted to (at most) the lifetime of the guard, and so it shouldn't be possible to get a reference with a longer lifetime than the guard that the reference was gotten from.

Reply:
yes, I agree that the borrow checker will do the right thing, it is more for the programmers sanity I am talking

Reply:
I don't think it's a good idea to introduce this kind of complexity into types that do not need to restrict lifetime anymore than the borrow checker already does—the `CacheState` type does not know that it is used in any way where the lifetime is not straight forward, nor does it contain any shared state like `Mutex` or `LocalKey` does.

That being said, if you *do* chose to go this route, I'd recommend at least accepting a return value for the closure:

```suggestion
    pub fn str_features<C, R>(&self, f: C) -> R
    where
        C: FnOnce(&HashMap<String, CachedFeature>) -> R,
    {
        f(&self.str_features)
    }
```

Reply:
I think thats a good point, but I need to halt() and sleep; I'll probably do some experiments without the closure to see what the errors look like when someone does something silly with it, and then either add R, or remove the closure. Thank you for the discussion :).

Reply:
So the errors are actually quite nice:

```rust
error[E0597]: `cache` does not live long enough
   --> src/main.rs:64:15
    |
62  |     let toggles = {
    |         ------- borrow later stored here
63  |         let cache = client.cached_state();
64  |         match cache.as_ref() {
    |               ^^^^^ borrowed value does not live long enough
...
120 |     };
    |     - `cache` dropped here while still borrowed
```

I'm still torn but am going to remove the complexity for now, as I know that extra complexity is a weakness I have.